cmd/process-agent/collector.go

```diff
@@ -128,91 +121,80 @@ func (l *Collector) run(exit chan bool) error {
 
-	if err := l.podForwarder.Start(); err != nil {
-		return fmt.Errorf("error starting pod forwarder: %s", err)
-	}
+	processPayloads := make(chan checkPayload, l.cfg.QueueSize)
+	podPayloads := make(chan checkPayload, l.cfg.QueueSize)
 
-	go util.HandleSignals(exit)
-	heartbeat := time.NewTicker(15 * time.Second)
-	queueSizeTicker := time.NewTicker(10 * time.Second)
+	var wg sync.WaitGroup
+
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
+
+		heartbeat := time.NewTicker(15 * time.Second)
+		defer heartbeat.Stop()
+
+		queueSizeTicker := time.NewTicker(10 * time.Second)
+		defer queueSizeTicker.Stop()
+
 		tags := []string{
 			fmt.Sprintf("version:%s", Version),
 			fmt.Sprintf("revision:%s", GitCommit),
 		}
 		for {
 			select {
-			case payload := <-l.send:
-				if len(l.send) >= l.cfg.QueueSize {
-					log.Info("Expiring payload from in-memory queue.")
-					// Limit number of items kept in memory while we wait.
-					<-l.send
-				}
-
-				for _, m := range payload.messages {
-					extraHeaders := make(http.Header)
-					extraHeaders.Set(api.HostHeader, l.cfg.HostName)
-					extraHeaders.Set(api.ProcessVersionHeader, Version)
-					extraHeaders.Set(api.ContainerCountHeader, strconv.Itoa(getContainerCount(m)))
-
-					if cid, err := clustername.GetClusterID(); err == nil && cid != "" {
-						extraHeaders.Set(api.ClusterIDHeader, cid)
-					}
-
-					body, err := encodePayload(m)
-					if err != nil {
-						log.Errorf("Unable to encode message: %s", err)
-						continue
-					}
-
-					payloads := forwarder.Payloads{&body}
-					var responses chan forwarder.Response
-
-					switch payload.name {
-					case checks.Process.Name():
-						responses, err = l.forwarder.SubmitProcessChecks(payloads, extraHeaders)
-					case checks.RTProcess.Name():
-						responses, err = l.forwarder.SubmitRTProcessChecks(payloads, extraHeaders)
-					case checks.Container.Name():
-						responses, err = l.forwarder.SubmitContainerChecks(payloads, extraHeaders)
-					case checks.RTContainer.Name():
-						responses, err = l.forwarder.SubmitRTContainerChecks(payloads, extraHeaders)
-					case checks.Connections.Name():
-						responses, err = l.forwarder.SubmitConnectionChecks(payloads, extraHeaders)
-					case checks.Pod.Name():
-						responses, err = l.podForwarder.SubmitPodChecks(payloads, extraHeaders)
-					default:
-						err = fmt.Errorf("unsupported payload type: %s", payload.name)
-					}
-
-					if err != nil {
-						log.Errorf("Unable to submit payload: %s", err)
-						continue
-					}
-
-					if statuses := readResponseStatuses(responses); len(statuses) > 0 {
-						l.updateStatus(statuses)
-					}
-				}
 			case <-heartbeat.C:
 				statsd.Client.Gauge("datadog.process.agent", 1, tags, 1)
 			case <-queueSizeTicker.C:
-				updateQueueSize(l.send)
+				updateQueueSize(len(processPayloads), len(podPayloads))
 			case <-exit:
 				return
 			}
 		}
 	}()
 
+	processForwarderOpts := forwarder.NewOptions(keysPerDomains(l.cfg.APIEndpoints))
```

Comment:
nit: things look a little repetitive here, any chance to make a for loop and simplify a little?

Reply:
What do you suggest?  I tried a few things but nothing that really felt cleaner or substantially reduced the LOC.

Reply:
I was gonna suggest having a map with key as forwarder objects and value as the payload channels but that's super lame, maybe let's just leave them :)